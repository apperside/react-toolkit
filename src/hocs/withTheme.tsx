import React from 'react';
import { baseTheme, ThemeProvider } from '../ui/theme';

export function withTheme<P>(WrappedComponent: React.ComponentType<P>): React.FC<P> {
    const ComponentWithTheme = (props: P) => {
        return (
            <ThemeProvider theme={baseTheme as any}>
                <WrappedComponent {...props} />
            </ThemeProvider>
        );
    };
    return ComponentWithTheme;
}
export default withTheme;
