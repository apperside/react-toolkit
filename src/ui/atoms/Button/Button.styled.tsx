import { PropsOf } from "src/typeUtils";
import styled from 'styled-components';
import {
    border, BorderProps, borderRadius,
    buttonStyle,
    color,
    ColorProps, flexbox, FlexboxProps, fontSize, layout, LayoutProps, padding,
    position, PositionProps, space, SpaceProps, typography, variant
} from 'styled-system';
import { CursorProps, PropsWithVariant } from "../../theme";
import { AssetName } from "../Asset";
import Text from "../Text/Text";
import { ButtonVariants } from "./Button.theme";

export type ButtonStyleProps =
    ColorProps &
    BorderProps &
    SpaceProps &
    LayoutProps &
    FlexboxProps &
    PositionProps &
    CursorProps &
    { textStyle?: PropsOf<typeof Text> }


export interface ButtonProps extends ButtonStyleProps, PropsWithVariant<keyof ButtonVariants> {
    // children: React.ReactNode;
    label?: string
    onClick?: () => void
    isLoading?: boolean
    icon?: AssetName | React.ReactNode
    iconPosition?: 'left' | 'right'
    iconSize?: number
}

export const ThemedButton = styled('button')<ButtonProps>(
    {
        appearance: 'button',
        border: 0,
        outline: 0,
    },
    borderRadius,
    // buttonSize,
    buttonStyle,
    color,
    space,
    fontSize,
    color,
    space,
    layout,
    padding,
    flexbox,
    border,
    typography,
    position,
    props => {
        return variant({
            variants: props.theme.button.variants as any
        })
    },

);
