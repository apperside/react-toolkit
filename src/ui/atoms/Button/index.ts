export * from "./Button"
export * from "./Button.styled"
export * from "./Button.theme"