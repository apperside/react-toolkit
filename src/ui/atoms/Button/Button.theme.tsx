import { PropsOf } from "src/typeUtils";
import { VariantProps } from "../../theme";
import { ButtonStyleProps } from "./Button.styled";

export interface ButtonVariants {
  primary: VariantProps<ButtonStyleProps>
  secondary: VariantProps<ButtonStyleProps>
  danger: VariantProps<ButtonStyleProps>
}

export type ButtonTheme = ButtonStyleProps & {
  variants: Partial<ButtonVariants>
}

export const buttonTheme: ButtonTheme = {
  borderWidth: 0,
  textStyle: {
    textAlign: "center",
    fontSize: "medium",
    color: "white",
  },
  px: 20,
  py: 5,
  borderRadius: 5,
  variants: {
    primary: {
      color: "white",
      backgroundColor: "primary",
    },
    secondary: {
      color: "white",
      bg: "secondary",
    },
    danger: {
      backgroundColor: "warning",
    }
  }
};