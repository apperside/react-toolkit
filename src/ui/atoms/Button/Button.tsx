import React, { useMemo } from 'react';
import { PropsOf } from "../../../typeUtils";
import { useTheme } from "styled-components";
import { ButtonProps, ThemedButton } from './Button.styled';

/**
 * Button UI component for user interaction
 */
export const Button: React.FC<PropsOf<typeof ThemedButton> & ButtonProps> = ({
    variant = "primary",
    label,
    ...props
}) => {
    const theme = useTheme();
    const buttonTheme = theme.button;
    const baseTheme = useMemo(() => {
        const { variants, ...restOfTheme } = buttonTheme;
        return restOfTheme;
    }, [buttonTheme])
    console.log("base theme is", baseTheme)
    return (
        <ThemedButton {...baseTheme} variant={variant} {...props} >
            {label ?? props.children}
        </ThemedButton >
    )
};



export default Button;
