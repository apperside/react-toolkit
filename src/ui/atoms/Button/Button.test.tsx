/* eslint-disable testing-library/no-node-access */
import React from "react";
import { render, screen } from "@testing-library/react";
import Button from "./Button";
import userEvent from "@testing-library/user-event";
import withTheme from "src/hocs/withTheme";

const ThemedButton = withTheme(Button);
describe("Button", () => {
  test("renders a default button with text", () => {
    const { baseElement } = render(
      <ThemedButton id="mybutton" label={"Click me"} />
    );
    const button = baseElement.querySelector("#mybutton");
    // const text = screen.getByText('Click Me');
    expect(button).toBeInTheDocument();
    // expect(text).toBeInTheDocument();
    expect(button).toHaveStyle({
      backgroundColor: "primary",
      color: "white",
    });
    // expect(text).toHaveStyle({
    //     color: 'white',
    //     fontSize: "1.125rem"
    // });
    expect(baseElement).toMatchSnapshot();
  });
  test("renders a primary button", () => {
    const { baseElement } = render(
      <ThemedButton id="mybutton" variant="primary" label={"Click me"} />
    );
    const button = baseElement.querySelector("#mybutton");
    // const text = screen.getByText('Click Me');
    // console.log("button style is", button.style)
    expect(button).toHaveStyle({
      backgroundColor: "primary",
      color: "white",
    });
    // expect(text).toHaveStyle({
    //     backgroundColor: 'primary',
    //     color: 'white',
    // });
    expect(baseElement).toMatchSnapshot();
  });

  test("renders a primary overriden button", () => {
    const { baseElement } = render(
      <ThemedButton
        id="mybutton"
        variant="primary"
        backgroundColor="blue"
        label={"Click me"}
      />
    );
    const button = baseElement.querySelector("#mybutton");
    // const text = screen.getByText('Click Me');
    // console.log("button style is", button.style)
    expect(button).toHaveStyle({
      backgroundColor: "blue",
      color: "white",
    });
    // expect(text).toHaveStyle({
    //     backgroundColor: 'primary',
    //     color: 'white',
    // });
    expect(baseElement).toMatchSnapshot();
  });
  test("handles onClick", () => {
    const mockOnClick = jest.fn();
    const { baseElement } = render(
      <ThemedButton id="mybutton" label={"Click Me"} onClick={mockOnClick} />
    );
    const button = baseElement.querySelector("#mybutton");
    userEvent.click(button!);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
    expect(baseElement).toMatchSnapshot();
  });
});

