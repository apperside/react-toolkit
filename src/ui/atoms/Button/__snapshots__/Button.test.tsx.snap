// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button handles onClick 1`] = `
<body>
  <div>
     
    <button
      class="sc-bdnxRM cOQJuI"
      id="mybutton"
    >
      <span
        class="sc-hKFxyN hHAztA"
        color="white"
        font-size="medium"
      >
        Click Me
      </span>
    </button>
     
  </div>
</body>
`;

exports[`Button renders a default button with text 1`] = `
<body>
  <div>
     
    <button
      class="sc-bdnxRM cOQJuI"
      id="mybutton"
    >
      <span
        class="sc-hKFxyN hHAztA"
        color="white"
        font-size="medium"
      >
        Click me
      </span>
    </button>
     
  </div>
</body>
`;

exports[`Button renders a primary button 1`] = `
<body>
  <div>
     
    <button
      class="sc-bdnxRM cOQJuI"
      id="mybutton"
    >
      <span
        class="sc-hKFxyN hHAztA"
        color="white"
        font-size="medium"
      >
        Click me
      </span>
    </button>
     
  </div>
</body>
`;
